### ISSUE TITLE
Building lua modules by loading `Lua` from existing state in a luaopen_mylib function

### ISSUE THREAD

**Issue by cormacrelf** [mlua-rs/rlua#105]
Context: I'm looking at building a Pandoc lua filter in Rust.

It would be great if rlua could run inside a pre-existing Lua state, and help with library creation.

Hlua doesn't seem to have this any more: if you translate the old [proc macro code](https://github.com/tomaka/hlua/blob/master/rust-hl-lua-modules/src/rust-hl-lua-modules.rs) into the output it would produce if it compiled, `lua.load_new_table()` is not a real function. It looks like you can only manipulate named variables, not the stack (correct me if I'm wrong here). Currently, it looks like the only way to do that is with the [rust-lua53 raw bindings](https://github.com/jcmoyer/rust-lua53/blob/master/examples/as-lua-lib/mathx.rs), if you can stomach the unsafety. I'm not sure I can, because my code is not the kind you operate yourself, it's pretty end-user-facing, and it might get pretty complicated.

Ideally, you would run `Lua::from_existing_state(lua, false)` like you can in hlua, and there would be a convenient API for doing what `luaL_newlib` does. Is this easy? Would it be as simple as replacing [this line](https://github.com/kyren/rlua/blob/master/src/lua.rs#L154) with an argument passed in?

**Comment by kyren** [comment]
I think it might be reasonable to allow an `rlua::Lua` to be created from an existing state, but it would *definitely* be unsafe.  And not just that it would require a `*mut ffi::lua_State` either, it would be more deeply unsafe than that.

rlua makes a lot of *assumptions* about the underlying Lua state that have to remain true to keep safety, things like C side metatables being correct that you can't change in Lua scripts but you *can* change with arbitrary C API access.  Either giving raw `*mut ffi::lua_State` access or allowing one to be passed in would come with a large list of caveats about what you can and can't do, and I'd have to think for a while to be able to list them all.  It might be that all of these invariants are reasonable and surface level, or there might be some real trickiness hiding there, I'm not quite sure yet.

There are some weird invariants that rlua keeps right now.  Just as an example, there's a really terrible "trick" that rlua has to do where it temporarily turns off garbage collection so that it can know that 'm' erroring functions in the C API don't longjmp.  I really don't actually like this trick and it's on my short term list of things to do to get rid of it, but it exists right now.  This is safe only because I control the allocator and know that the allocator aborts instead of ever erroring, but it would become very unsafe if you were able to replace the global Lua allocator, or rlua wasn't allowed to replace the allocator when handed an "external" Lua state pointer.  It would probably work 99.999% of the time, which in a way is the *worst* kind of unsafety: it isn't a loud problem but is instead a vulnerability waiting to be discovered.  This probably isn't the only example, but it's the first one that came to mind, and It's these kind of things that made me decide *not* to hand out the state pointer to begin with.  Now, I think in the near term I'm going to get rid of this `gc_guard` nonsense, and that will remove the requirement that rlua control the allocator function, but there are a lot of other things *like* this.  Basically, every complication that the debug library has plus a bunch more.

Given all that, I still think it would be a reasonable feature to add at least *after* I remove the requirement to control the allocator function.  It feels a bit "rude" to me to allow you to use an external state pointer but require that I be given the ability to override the global state pointer.  I might have to think a bit more though, because another very "global" thing that rlua is about to allow you to do is set and clear the global hook function, but at least in that case it will be entirely user controlled.

I think the whole business with `luaL_newlib` is a lot simpler, it seems like it's just a convenience for creating a table?  I'm not entirely sure it's even necessary to add if it's that simple, does `luaL_newlib` do something else magical that I'm not seeing?

**Comment by kyren** [comment]
The caveat list is basically a bunch of rules around what you can and can't do with the C API, it's not at all expressible with a trait, sorry.

LuaJIT's FFI library is cool, but it's kind of a different set of functionality entirely, right?  I don't think LuaJIT's FFI library lets you write C code that interacts with Lua at all, correct?  Lua can *call* the C code, but you couldn't for example write a C function that inserts into a Lua table and call that using the FFI library, so C code in this case doesn't even see the Lua state pointer.  Maybe I'm misunderstanding your use case here?

Like you said, it's really not much work to actually add the functionality, so after I remove the custom allocator requirement I can look at adding it and *lightly* document it.  I shouldn't really be so worried, because this rabbit hole already exists with the debug library, it is possible to segfault the interpreter using `debug` independently of rlua with just the Lua C library and a script, so it's hopelessly unsafe, this wouldn't really be much worse.  I'm fine with marking them in rlua with the unsafe keyword and just assuming you know what you're doing, I'll just try and give *examples* of how tricky they can be so people understand the danger.

**Comment by cormacrelf** [comment]
Ah, ok. My use case really is just calling Rust functions with some strings and structs, and reading the resulting structs. Using LuaJIT FFI would just mean using structs directly and converting back to tables in Lua, and almost zero other work.

**Comment by kyren** [comment]
So it sounds like you're trying to place rust + rlua in the *library* position of Lua code rather than the main application, right?  I think I was focused a bit too much on the idea of the user controlling the state pointer and not this use case in general, and I don't think I thought it through enough.  If `rlua` is going to be used to make Lua libraries, I need to start thinking about things like: what happens when there are multiple `rlua` based Lua libraries in use?  What if they use different versions of `rlua`?  Of rustc?  There might be a few more difficulties than I originally thought.

There is quite a bit of "setup" that must happen to the Lua state before `rlua` can use it at all.  Obviously right now `rlua` loads packages like 'io' and 'coroutine' on startup and it definitely shouldn't do this when adopting an external pointer, but there are some things that setup *must* do like creating metatables for rust function and error types.  These are placed in the registry with keys based on the address of static variables, and this will actually work fine with multiple `rlua` based libraries in play (each linked version of `rlua` will get its own static variables with unique addresses).  It may add some small startup time, but this is workable in "library position".

One problem though is the fact that `rlua` wraps the global `pcall` and `xpcall` functions to ensure that panics cannot be captured by Lua, and this is definitely *not* workable in library position.  Besides the fact that this is another "rude" global change, it would also only work for the *last* `rlua` library to run the setup procedure.  Luckily, this is not strictly a safety issue as it "only" lets Lua swallow Rust panics, so I think this is fine to selectively disable, and disabling it is the right choice for being in library position anyway.

If those are the only problems then actually it *does* still sound workable to make `rlua` based Lua libraries, but there might be something else I missed.

Regarding LuaJIT FFI, yeah `rlua` has a long road to go before it is as convenient as that.  I think the proper alternative to LuaJIT FFI might not be something like `rlua` itself but rather, something like `cbindgen` but for Lua.  The task is in some ways harder, but actually it is in *many* ways easier, since this kind of API is not a bidirectional one, and that saves a *ton* of complexity.  In fact, you can kind of see this in LuaJIT: it still has the normal C API and requires its use, but the basic advice is "it's slow and complex, try not to use it very much".  Maybe in the future when the semantics and safety parts of `rlua` are settled and the basic API is settled some more work can be done to try to get to the level of convenience of the LuaJIT FFI library, even though `rlua` is a more complex fully bidirectional one.

**Comment by cormacrelf** [comment]
Well, I'm not blocked on this, because Pandoc has a perfectly good JSON-serialized intermediate format to use instead. So take all the time you need, or don't do it at all! But this is certainly a challenge that only Rust is really up to, and rlua is in a unique position to make it safer.

**Comment by kyren** [comment]
I definitely won't do "cbindgen but for Lua" soon, but I'll probably think about properly supporting rlua being used in library position, compiled into a shared object and things like that.

**Comment by kyren** [comment]
One change that will need to be made to support `rlua` in library position is not storing `ExtraData` in the extra data region of the main `lua_State`, or at least not storing anything there that would be used when `rlua` is used in this way.  This is not especially hard since the registry exists, but it *is* potentially slow.

**Comment by khvzak** [comment]
[mlua](https://github.com/khvzak/mlua) supports writing Lua modules in Rust. It does not store anything in `lua_State` and keeps the global `pcall/xpcall` functions unchanged (like any other).
